### ISSUE TITLE
structs' public fields should be before private fields (and do we actually need pub keyword?)

### ISSUE THREAD

**Issue by vladimir-kraus** [vlang/v#6582]
Readability matters and I find V to be very readable overall. Except one thing. When looking at struct definition we always see private fields first and public fields only after them in section delimited by "pub:". Should not this be somehow inverted, seeing public items first should help improve readability greatly.

Of course I can mark each public field with "pub" individually and list them at the top of struct definition, but that is very verbose.

UPDATE: Crucial question is whether distinguishing private and public fields adds any significant value for the language users. Some popular modern languages like Python or Kotlin have all their fields public by default (unless marked as private). I believe that leaving this symbol visibility feature out of the language definition would simplify the language very much without sacrificing anything significant. Of course it would then be up to the user to decide to communicate that some parts of struct is private and not intended to be touched by caller, e.g. by underscoring the field name like in Python. Enforcing this convention would then be up to static analyzers instead of the compiler. But the language itself would remain clean and simple.

**Comment by JalonSolov** [comment]
I find it easier to notice the `pub:` if it is NOT on the line immediately under the `struct` line.

**Comment by vladimir-kraus** [comment]
But if the `struct` has only public fields, then `pub:` actually is immediately under the `struct` line. Unless you leave a blank like between `struct` and `pub:`.

**Comment by JalonSolov** [comment]
Yes, and I'm fine with it being immediately under if there are ONLY pub fields.  If there are a mix of pub and non-pub, I prefer to see the separation.

**Comment by vladimir-kraus** [comment]
Another question is whether there is a really widespread use for private fields. Most of the structs I have seen in the codebase have more public fields than private fields, often no private fields at all. This means that we almost always have to write `pub` somewhere.

For example in Python this is solved by one or two underscores to denote the private ones. All others are public. I find this quite handy.

**Comment by vladimir-kraus** [comment]
Another question is whether private fields are actually useful. Maybe I am wrong but the only purpose is to isolate the user of the `struct` instance from implementation details and allow him to use only the public interface. Which means that there MUST be a public interface (otherwise the struct instance is of no use) and CAN be private details. 

The philosophical question should be whether more typing (I mean typing the extra three letters "p", "u", "b") should support the major use case or whether it should enforce API safety and proper use. In my opinion less typing should be pragmatic and support the major use case, which is the public visibility. I see V as simple and pragmatic, not as super-safe (it is no Rust and never will be). In that case requiring `pub` to mark public fields goes against pragmatism and favors safety.

For example in Kotlin all fields are public unless marked as private. While in Java it is the opposite. Needless to say that Kotlin's approach is more pragmatic and popular among programmers than old-school Java's approach. And V goes the Java way.

**Comment by ntrel** [comment]
> seeing public items first should help improve readability greatly

I also think that would be better, private fields are usually of less interest. You can do it with `mut:` but immutable private isn't supported AFAIK. Perhaps a `default` modifier would work:
```v
struct S1 {
pub:
  visible T
default:
  secret T
}
```

> Which means that there always MUST be a public interface (otherwise the struct instance is of no use)

You can have public methods without having any public fields. This is a very important concept e.g. for ensuring memory safety. Reading and copying fields may not even be memory-safe for certain structs - but they can provide methods that return memory-safe values.

> The philosophical question should be whether more typing (I mean typing the extra three letters "p", "u", "b") should support the major use case

More typing should be necessary when giving away power to user code. You will break user code if you later change a public field to private. `pub` by default might mean accidentally public fields that you later regret. (Also symbols in a module are private by default, so this is consistent with that).

> I see V as simple and pragmatic, not as super-safe (it is no Rust and never will be)

V's goal is 100% memory-safety (outside `unsafe {}`), just like Rust. V will support that in a simpler way than Rust (which should be less limiting as it doesn't require a unique reference to mutate shared data).

**Comment by vladimir-kraus** [comment]
Yes, I agree with what you write. I was mostly concerned about the clash between pragmatism and safety. When I look at V's standard library, I see there are maybe more than 90 percent of all functions and structs and struct fields marked as public (at least it was my impression, I was not counting it exactly). And seeing `pub` everywhere is just so much visual clutter. However if V is supposed to aim more safety than pragmatism, than using `pub` to explicitly denote public stuff is the obvious strategy. I am fully OK with that.

Regarding the consistency of visibility between functions and structs vs struct fields. I do not think we must enforce full consistency, they are already not consistent anyway. Because for struct fields you can write `pub:` to mark a whole group of public fields. You cannot do this with functions and structs (AFAIK). Which in turn means that you may add a field which you intend to be private to this public group by mistake if you fail to spot `pub:` a few lines (maybe in some cases it is many lines) above...

Which means that `pub` and `pub:` are actually syntactically/semantically very different beasts. Maybe we should not unite them under one keyword. Maybe `public:` would be better to mark the whole group. Due to the fact that it is two times longer, it is easier not to overlook and it will not be mistaken with a single item qualifier `pub`. And when we introduce `public:` then there is no problem with introduction of `private:` as well. Which will give us the option to put public fields before private fields.

**Comment by vladimir-kraus** [comment]
And I feel there will be another issue with embeded structs.

    struct Button {
        Widget
        title string
    }

Are Widget's fields public or private when accessed via Button? Maybe it is a wrong question because I guess the intention is to inherit the visibility of `Widget` fields to `Button`. But now `Widget` is listed among private fields which are always at the top. At least visually when reading the code this feels very confusing.

**Comment by JalonSolov** [comment]
Also consider modules vs your main code.  There are many reasons to not mark fields as pub in modules (as they are private to the module itself), so yes the `pub` keyword is needed.

Since it is `pub:` to mark a section from there to the bottom of the struct, it does need to be _after_ the private fields.

As far as embedded structs, yes, the visibility should be inherited.

**Comment by vladimir-kraus** [comment]
Oh, I did not know that fields which are not marked as pub are private for the module itself. Well, in that case this is yet another inconsistency between `pub` for functions and structs and `pub:` for fields. Using one keyword for such different rules could be a kind of confusing. It really gives wrong impression about the consequences and its usage.

**Comment by JalonSolov** [comment]
pub means pub.  If you don't mark a function as pub, it is private to the module it's in.  If you don't mark a struct pub, it is private to the module it's in.  There is no inconsistency.

Also 'private' in V is similar to 'protected' in Java (if you're familiar with that).  Meaning it is still accessible to the whole module (package), but private to anything outside.

**Comment by vladimir-kraus** [comment]
My fault, I misunderstood what you meant by private fields.

**Comment by medvednikov** [comment]
I agree, public fields should be on top.
